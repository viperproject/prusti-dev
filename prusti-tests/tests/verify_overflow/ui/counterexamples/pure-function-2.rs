// compile-flags: -Punsafe_core_proof=true -Pcounterexample=true -Pcheck_overflows=false

use prusti_contracts::*;

#[derive(Copy, Clone)]
struct X {
    a: i32
}

impl X {
    #[pure]
    #[terminates(Int::new(1))]
    fn get_a(&self) -> i32{
        self.a
    }
}


#[pure]
#[terminates(Int::new(1))]
fn baz(x: X) -> X {
    X {a: x.a + 1}
}

#[requires(y.a == 5)] //force specific counterexample
fn test(x: i32, y: X) {
    //FIXME: Counterexample generation treats y.get_a() as an assignment for y and
    //produces a counterexample of y at that point
    let z = y.get_a();
    assert!(z == baz(y).a)
}


fn main(){}
