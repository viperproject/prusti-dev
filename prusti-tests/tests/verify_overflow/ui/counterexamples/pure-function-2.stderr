error: [Prusti: verification error] the asserted expression might not hold
  --> $DIR/pure-function-2.rs:30:5
   |
30 |     assert!(z == baz(y).a)
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
note: the failing assertion is here
  --> $DIR/pure-function-2.rs:30:5
   |
30 |     assert!(z == baz(y).a)
   |     ^^^^^^^^^^^^^^^^^^^^^^
note: counterexample for "x"
       value:   2
  --> $DIR/pure-function-2.rs:26:9
   |
26 | fn test(x: i32, y: X) {
   |         ^
note: counterexample for "y"
       value:   X {
            a: 5,
        }
  --> $DIR/pure-function-2.rs:26:17
   |
26 | fn test(x: i32, y: X) {
   |                 ^
note: counterexample for "y"
       value:   X {
            a: 5,
        }
  --> $DIR/pure-function-2.rs:29:13
   |
29 |     let z = y.get_a();
   |             ^^^^^^^^^
note: counterexample for "z"
       value:   5
  --> $DIR/pure-function-2.rs:29:13
   |
29 |     let z = y.get_a();
   |             ^^^^^^^^^
note: counterexample for "get_a()"
       value:   5
  --> $DIR/pure-function-2.rs:29:13
   |
29 |     let z = y.get_a();
   |             ^^^^^^^^^
note: counterexample for "baz()"
       value:   X {
            a: 6,
        }
  --> $DIR/pure-function-2.rs:30:18
   |
30 |     assert!(z == baz(y).a)
   |                  ^^^^^^
   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

